src/basic/virt.c

```diff
@@ -230,13 +230,26 @@ static int detect_vm_smbios(void) {
 static int detect_vm_dmi(void) {
 #if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || defined(__aarch64__)
 
-        int r;
+        int r, r1;
+        _cleanup_free_ char *s = NULL;
+        size_t readsize;
         r = detect_vm_dmi_vendor();
 
         /* The DMI vendor tables in /sys/class/dmi/id don't help us distinguish between Amazon EC2
          * virtual machines and bare-metal instances, so we need to look at SMBIOS. */
-        if (r == VIRTUALIZATION_AMAZON && detect_vm_smbios() == SMBIOS_VM_BIT_UNSET)
-                return VIRTUALIZATION_NONE;
+        if (r == VIRTUALIZATION_AMAZON) {
+                switch (detect_vm_smbios()) {
+                case SMBIOS_VM_BIT_UNKNOWN:
+                        r1 = read_full_virtual_file("/sys/class/dmi/id/product_name", &s, &readsize);
+                        if (r1 < 0 || !strstr(s, ".metal"))
+                                break;
```

Comment:
please add a comment here explaining why we fall back to reading the product name. Also, maybe provide an example string that amazon systems expose here IRL.

Also, is strstr() really the right thing? i.e. shouldn't it be endswith()? And if not endswith(), then what comes after the string? isn't there any form of delimiter after it?

Reply:
Ok will do. As for strstr ... it was easy but yes, endswith() might be better (I hadn't noticed we had that helper). There should be nothing else. I'll double check our instance names catalog to make sure that's right before I submit an update.

Reply:
So it looks like endswith() won't do because the string has a \n in it, and I don't want to bet that this is consistent  accross firwmare variants accross the fleet. So I'll stick to strstr for now

Reply:
Bah ignore this... it's the kernel who is adding \n ...